projects/igniteui-angular/src/lib/calendar/calendar.ts

```diff
@@ -388,12 +388,30 @@ export class Calendar {
         return this.timedelta(date, TimeDeltaInterval.Year, -1);
     }
 
-    public getWeekNumber(date: Date) {
-        const firstJan = new Date(date.getFullYear(), 0, 1).getTime();
-        const today = new Date(date.getFullYear(), date.getMonth(), date.getDate()).getTime();
-        const dayInMilSeconds = 86400000;
-        const dayOfYear = ((today - firstJan + 1) / dayInMilSeconds);
-        return Math.ceil(dayOfYear / 7);
+    public getWeekNumber(date: Date, weekStart: WEEKDAYS | number) {
+        const yearStart = new Date(date.getFullYear(), 0, 1);
+        //the day of the week on which the year begins
+        let firstDayOfTheYear = yearStart.getDay() - weekStart;
+        firstDayOfTheYear = firstDayOfTheYear >= 0 ? firstDayOfTheYear : firstDayOfTheYear + 7;
+        const dayInSeconds = 86400000;
+        const dayNumber = Math.floor((date.getTime() - yearStart.getTime() - 
+            (date.getTimezoneOffset() - yearStart.getTimezoneOffset()) * 60000) / dayInSeconds) + 1;
+        let weekNumber;
+        //if the year starts before the middle of a week
```

Comment:
Maybe put more comments in the whole function, that make it clear what the logic is:

for example - if 01 Jan is Monday to Thursday, is considered 1st week of the year and vice -versa - if 01 Jan starts Friday to Su, is considered last week of previous year.